File: 2019/levelUp302/lecture_002_Graph/l003_UnionFind.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
class Edge
{
public:
    int v = 0;
    int w = 0;

    Edge(int v, int w)
    {
        this->v = v;
        this->w = w;
    }
};

int N = 7;
vector<vector<Edge>> graph(N, vector<Edge>());

void addEdge(vector<vector<Edge>> &gp, int u, int v, int w)
{
    gp[u].push_back(Edge(v, w));
    gp[v].push_back(Edge(u, w));
}

//unionFind.==============================================

vector<int> par;
vector<int> setSize;

int findPar(int vtx)
{
    if(par[vtx]==vtx) return vtx;
    return par[vtx]=findPar(par[vtx]);
}

void mergeSet(int l1, int l2)
{
    if (setSize[l1] < setSize[l2])
    {
        par[l1] = l2;
        setSize[l2] += setSize[l1];
    }
    else
    {
        par[l2] = l1;
        setSize[l1] += setSize[l2];
    }
}

void unionFind(){

}



void constructGraph()
{
    addEdge(graph, 0, 1, 10);
    addEdge(graph, 0, 3, 10);
    addEdge(graph, 1, 2, 10);
    addEdge(graph, 2, 3, 40);
    addEdge(graph, 3, 4, 2);
    addEdge(graph, 4, 5, 2);
    addEdge(graph, 4, 6, 3);
    addEdge(graph, 5, 6, 8);

    // addEdge(graph, 2, 5, 2);

    display(graph);
    cout << endl;
}

void solve()
{
    constructGraph();
}

int main()
{
    solve();
    return 0;
}

File: 2019/BPIT_PP_Batch/lecture_006_Graph/l003_unionFind.java
public class l003_unionFind{
     
    public static class Edge{
        int v = 0;
        int w = 0;

        Edge(int v,int w){
            this.v = v;
            this.w = w;
        }
    }

    static int N = 7;
    static ArrayList<Edge>[] graph = new ArrayList[N];

    public static void addEdge(int u,int v,int w){
        graph[u].add(new Edge(v,w));
        graph[v].add(new Edge(u,w));
    }

    public static void display(){
        StringBuilder sb = new StringBuilder();
        for(int i=0;i < N;i++){
            sb.append(i + " -> ");
            for(Edge e: graph[i]){
                sb.append("(" + e.v + ", " + e.w +") ");
            }
            sb.append('\n');
        }

        sb.append('\n');
        System.out.println(sb.toString());
    }

    //=======================================================================

    static int[] par;
    static int[] size;

    public static int findPar(int u){
        if(u == par[u]) return u;
        return par[u] = findPar(par[u]);
    }

    public static void merge(int p1,int p2){
        if(size[p1] < size[p2]){
            par[p1] = p2;
            size[p2] += size[p1];
        }else{
            par[p2] = p1;
            size[p1] += size[p2];
        }

    }

    // {{u,v,w}}
    public static void unionFind(int[][] gp){  
        for(int i=0;i<N;i++) graph[i] = new ArrayList<>();
        
        Arrays.sort(gp,(a,b)->{
            return a[2]-b[2]; // this - other,default, Increasing
            // return b[2]-a[2]; // other - this, reverse of default, Decreasing
        });

        for(int i=0;i<N;i++){
            par[i] = i;
        }

        for(int[] a: gp){
            int u = a[0];
            int v = a[1];
            int w = a[2];

            int p1 = findPar(u);
            int p2 = findPar(v);

            if(p1 != p2){
                addEdge(u,v,w);
                merge(p1,p2);
            }
        }
    }
}